index: Fix backwards search for bestblock

mzumsande opened the pull request:
This PR attempts to fix an intermittent Init issue encountered during the stress testing of #23289, which relates to the pruning-compatible filter reconstruction logic introduced in #15946. 

The problem would occur when the node starts with `-txindex=1` but `ThreadSync` is interrupted after it sets `m_best_block_index` to Genesis, and before it gets do any further work.
In that case, during the next restart of the node, an Init error would be thrown because  `BaseIndex::Init()` tries to backtrack from the tip to the last block which has been successfully indexed (here: Genesis), but the backtracking logic didn't work properly in this case: 
The loop 
`while (block_to_test && block->pprev && (block->pprev->nStatus & BLOCK_HAVE_DATA))` checks if a predecessor exists **before** performing the check `block_to_test == block` and then possbily setting `prune_violation = false` 
If `block_to_test` and `block` are the Genesis block this check will not be reached because `block->pprev` does not exist.

To reproduce this bug on regtest:
1) start a node with a fresh datadir using `-txindex=1` (or any other index)
2) stop and restart without any index
3) mine a block
3) stop and restart again with the index enabled 
->InitError `Error: txindex best block of the index goes beyond pruned data. (...)`

Fix this by requiring that we have the data for the block of the current iteration `block` (instead of requiring it for the predecessor `block->pprev`) 
That way, the check for `block_to_test == block` is also reached when `block_to_test` is the Genesis block.
No longer requiring the data of `block->pprev` also means that we can now prune up to `m_best_block_index` height without requiring a reindex (one block more than before). I added this edge case to `feature_blockfilterindex_prune.py`, the new version should fail on master.

mzumsande committed:
index: Fix backwards search for bestblock

This allows filters to be reconstructed when the best known block is
the Genesis block without needing to reindex.
It fixes Init errors seen in #23289.

mzumsande committed:
test: Add edge case of pruning up to index height

mzumsande commented:
Moved out of draft and added steps to reproduce to OP.
Not sure if the second commit (test) should be merged, maybe it's too detailed, plus the test gets reworked in #21726 anyway. 

Note that the last commit of #21726 would partly "fix" the issue too by just not running the affected code unless `-prune=1` is specified but doesn't address the root cause.

MarcoFalke commented on test/functional/feature_blockfilterindex_prune.py:
```diff
@@ -24,11 +24,7 @@ def run_test(self):
-        self.sync_all()
-        self.generate(self.nodes[0], 250)
-        self.sync_all()
+        self.generate(self.nodes[0], 500)
```
Why is this change correct?

mzumsande replied:
Because preventing disconnections due to NODE_NETWORK_LIMITED_MIN_BLOCKS makes little sense  when there is just one node involved in the test.

laanwj commented:
Concept ACK

jamesob reviewed (approved):
ACK a36e23b183a6b8d1c60598618f2a5c8f31da02e0 ([`jamesob/ackr/23365.1.mzumsande.index_fix_backwards_sear`](https://github.com/jamesob/bitcoin/tree/ackr/23365.1.mzumsande.index_fix_backwards_sear))

Thanks for looking into this and figuring out a good fix. Once this is merged, I can update #23289 to cover interrupting init after index initialization.

---

I built this locally and confirmed that the attached test fails if the `index/base.cpp` changes are reverted. Logically the change makes sense - if the index's best block is the genesis, we should be able to continue indexation from that point (which is all the rephrased `while` condition does).

The `block = block->pprev;` does make me a little nervous (although as-written the code would never cause any problems given how we initialize `block_to_test`), so if you wanted to you could include an assertion like 

```diff
diff --git a/src/index/base.cpp b/src/index/base.cpp
index 88c74ec543..8525dcbfa0 100644
--- a/src/index/base.cpp
+++ b/src/index/base.cpp
@@ -96,6 +96,7 @@ bool BaseIndex::Init()
                     prune_violation = false;
                     break;
                 }
+                assert(block->pprev);
                 block = block->pprev;
             }
         }
```

<details><summary>Show signature data</summary>
<p>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK a36e23b183a6b8d1c60598618f2a5c8f31da02e0 ([`jamesob/ackr/23365.1.mzumsande.index_fix_backwards_sear`](https://github.com/jamesob/bitcoin/tree/ackr/23365.1.mzumsande.index_fix_backwards_sear))

-----BEGIN PGP SIGNATURE-----

iQIzBAEBCgAdFiEEGNRVI1NPYuZCSIrGepNdrbLETwUFAmGbxWwACgkQepNdrbLE
TwVIdg/8CKYBVyw9DvvSx/G2E31/z7avW4uwj5cq10HMmjtsT/cCHhFtofDtVvIG
9binTF/dlYhyzQmv7FduzQkgpZtehokrrBugoktyh6zeJpRV+mPUxZvFJ0VUCK1w
+dVJ2zdbDmDuogJR7+ZZbGgBGxPKjS9t/Q+/LQIP221lgAVi1MPJ8oYuyg4+R6hn
ueS3piEtXmHnQt0F05PUhJDZZidB/3hRp1bu9Z0H79TrL65wnA1bnDf3+IKbGeQL
CPsYxZR4VZRiIqkSESsKuIbIVE2ELNJcQHP5pKoA389dBDbDlMfw1Lo/N+opB1tm
kbkrCt0ZB4HEAwsbyi3KWOue5rYkx5zuz6/7QW6sChkfGLwE+F+UfyvzCzy2P8FI
Fzd8lx/74O/+0NtcIpY+kSlPxztdQCuqCZHmx8cuYvnfPDIWkg2TrgX77JPoGJHO
x8mJBGWrTkxYoEM2brLWBKK24pPbafrx860NWEzJEoJ5xreZOfAzsBZPLtEZtymq
hggWvPjgGljGUFnKakjhP3ahMgDBNyQxoXM/n/XIE3NSepvRE7WOc5yxj2gGScM8
h6oIYZVmlgbbImA1asNDgXRk4acrA/N+3dZdUCAyjPz0lhx3/g2XznCZsfRyZdjA
cftewu0RF0l/XnyRZ37K5xvTUsjDYqb2omufwnmgy2giD9QnHdA=
=5ydW
-----END PGP SIGNATURE-----

```

</p></details>

<details><summary>Show platform data</summary>
<p>

```
Tested on Linux-5.10.0-9-amd64-x86_64-with-glibc2.28

Configured with ./configure LDFLAGS=-L/home/james/src/bitcoin/db4/lib/ CPPFLAGS=-I/home/james/src/bitcoin/db4/include/ CXXFLAGS=-fPIE -pipe -O2 -g -Wthread-safety-analysis -Wall -Werror=sign-compare -Wsign-compare -Werror=thread-safety-analysis --enable-wallet --enable-debug --with-daemon --enable-natpmp-default CXX=/usr/bin/clang++ CC=/usr/bin/clang PYTHONPATH= --disable-shared --with-pic --enable-benchmark=no --enable-module-recovery --enable-module-schnorrsig --enable-experimental --disable-openssl-tests --disable-shared --with-pic --enable-benchmark=no --enable-module-recovery --enable-module-schnorrsig --enable-experimental --disable-openssl-tests --no-create --no-recursion

Compiled with /usr/bin/ccache /usr/bin/clang++ -std=c++17 -mavx -mavx2 -fPIE -pipe -O2 -g -Wthread-safety-analysis -Wall -Werror=sign-compare -Wsign-compare -Werror=thread-safety-analysis -O0 -g3 -ftrapv -fdebug-prefix-map=$(abs_top_srcdir)=.  -Wstack-protector -fstack-protector-all -fcf-protection=full -fstack-clash-protection -msse4 -msha -msse4.1 -msse4.2  i

Compiler version: Debian clang version 13.0.1-++20211019122913+8a93745a7121-1~exp1~20211019003538.10
```

</p></details>

mzumsande commented:
> The `block = block->pprev;` does make me a little nervous (although as-written the code would never cause any problems given how we initialize `block_to_test`), so if you wanted to you could include an assertion like

Thanks - I added the assertion as suggested, and rebased. Note that adding the assert only changes the kind of error: Before, if `block->pprev` had not existed (which I agree shouldn't be possible), the InitError would have been thrown  (because the while condition includes a check for `block`).

jnewbery commented on src/index/base.cpp:
```diff
@@ -91,11 +91,12 @@ bool BaseIndex::Init()
                     prune_violation = false;
                     break;
                 }
+                assert(block->pprev);
```
It may be worth adding a comment here on _why_ `block->pprev` must exist (because we're walking backwards from the active chain tip to a block index in the active chain. If a block index that we encounter doesn't have a backwards pointer, then the chain is malformed).

jamesob replied:
This isn't actually a correct explanation (the genesis block is in the active chain and doesn't have `pprev` set). The reason this assertion will hold is because of how `block` and `block_to_test` are chosen; the conditional above this line is guaranteed to break at-or-higher-than genesis because of the few lines of logic above (which, if summarized in a comment, would just end up repeating the fairly-straightforward code above).

jnewbery replied:
I think you're saying the same thing as I am. Because of the way `block` and `block_to_test` are chosen, if we walk backwards from `block`, we're guaranteed to encounter `block_to_test` (either at or before the genesis block).

mzumsande replied:
Sorry I didn't get to this earlier, would have added a comment, do people think that this warrants a follow-up?

fanquake replied:
@mzumsande sounds like it could be.

mzumsande replied:
done in #23777

jnewbery commented on test/functional/feature_blockfilterindex_prune.py:
```diff
@@ -39,16 +37,29 @@ def run_test(self):
+        self.generate(self.nodes[0], 502)
+
+        self.log.info("prune exactly up to the blockfilterindexes best block while blockfilters are disabled")
+        pruneheight_2 = self.nodes[0].pruneblockchain(1000)
+        assert_equal(pruneheight_2, 998)
```
Is there a specific reason why calling `pruneblockchain(1000)` prunes up to exactly block 998? Is it because that's where the block file happens to wrap (i.e. blocks 999 and 1000 are in the same file as later blocks and so can't be pruned). If so, that seems a bit brittle (e.g. if the default block file size of disk serialization changes, then this test will fail), and definitely seems worthy of a comment.

mzumsande replied:
> Is it because that's where the block file happens to wrap

Yes, I believe that is the reason, and I agree it's brittle (although the test relied before on this, e.g. at `assert_equal(pruneheight, 248)` - I'm not sure though if there is a better way to test pruning in a detailed way.

mzumsande replied:
comment added in #23777

2310Mas replied:
Star

jnewbery reviewed (commented):
Concept ACK.

ryanofsky reviewed (commented):
Partial code review ACK 9600ea01450b0d39be90eb2971c1ac5c9b69a66e for the code change, not the test changes. (Test changes are indirect and little over my head.) It seems obvious that previous code `prune_violation = true, while (block->pprev)` would incorrectly detect a prune violation at the genesis block, and the fix here make sense and looks correct.
